lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -81,8 +81,29 @@ def add_node_data(node)
     node.add_server_facts(@server_facts)
   end
 
+  # Inline file metadata for static catalogs
+  # Initially restricted to files sourced from codedir via puppet:/// uri.
+  def inline_metadata(catalog, environment)
+    catalog.resources.find_all { |res| res.type == "File" }.each do |resource|
+      next if resource[:ensure] == 'absent'
+
+      next unless source = resource[:source]
+      next unless source =~ /^puppet:/
+
+      file = resource.to_ral
+
+      if file.recurse?
+        # TODO: find and replace metadata for children recursively
+      else
+        # TODO: find and replace metadata for the file
+      end
+    end
+
+    catalog
+  end
```

Comment:
Separating it out as a new class would make testing easier (`compile` is private so testing is indirect). Related, https://github.com/joshcooper/puppet/commit/81d4f41f2f19a8a371f9f5c13cd67b1005b24c3a has already tread this ground a bit.

I'll investigate whether we can inline resources in the `Compiler#filter`, or if that's too late.